### ISSUE TITLE
Writes to cartridge RAM using read port should fail

### ISSUE THREAD

**Issue by ale-79** [stella-emu/stella#537]
See here

[https://atariage.com/forums/topic/293315-sokoban/?do=findComment&comment=4347497](https://atariage.com/forums/topic/293315-sokoban/?do=findComment&comment=4347497)

On real hardware (Unocart), you get pulsating walls, while they're static in Stella.
I quickly checked on Stellerator, javatari and z26, and they match real hardware

**Comment by andrew-davie** [comment]
This is likely to be an unitialised variable in my code, with different default RAM contents. I have tried Stella with the randomised RAM on/off and see no difference, but it's probably luck of the draw. I will report back.

**Comment by andrew-davie** [comment]
I am still looking at this, but now sure it is an initialisation error. The only reason I am not closing it is because I want to understand how/why I see no different behaviour in Stella when I randomise RAM or not. I'll report back soon.

**Comment by andrew-davie** [comment]
OK, I've tracked this down.

This **is** a bug in Stella.

In bankswitch scheme 3E, when RAM is switched in ($F000-$F3FF) - 1K - you read from that address range, but if you want to write, you add $400 and write to that resultant address.  Thus,

```
Colour = $F200 . ; for example
   lda Colour . ; this will READ the byte at $F200
   sta Colour+$400  ; this will WRITE the byte at $F200 (not $F600)
```

That's how it's supposed to work, and it does.
But, in my code that wasn't working, I *forgot* that the variable was in switchable RAM, and thus left out the $400 offset when doing a write. Stella didn't care - it still changed the value in RAM correctly. This is not correct behaviour, and explains why the other emulators behaved differently.

Basically, the implementation of 3E seems to have a bug where writes to RAM *without* the $400 offset act exactly the same as writes to RAM *with* the offset.  That's my diagnosis.

**Comment by thrust26** [comment]
Maybe we should rename the issue then.

**Comment by andrew-davie** [comment]
> Maybe we should rename the issue then.

That would be fine by me, but I do not appear to have the power to do that.

**Comment by ale-79** [comment]
I changed the title to something more descriptive. 

BTW, maybe a develper option that breaks on writes to read-only addresses could be useful to track down these sort of bugs.

**Comment by sa666666** [comment]
We already can break on 'read to write port'.  I guess it wouldn't be too hard to catch the opposite too (write to read port).  Although on a real console this will never have any effect (unlike the former, where RAM could be screwed up).

**Comment by sa666666** [comment]
@thrust26, with your commits above, is this issue now solved?  IOW, can we close?

**Comment by thrust26** [comment]
I doubt my changes are complete.
